mesonbuild/compilers/compilers.py

```diff
@@ -689,14 +691,36 @@ def run(self, code: 'mesonlib.FileOrString', env: 'Environment', *,
             dependencies: T.Optional[T.List['Dependency']] = None) -> RunResult:
         raise EnvironmentException('Language %s does not support run checks.' % self.get_display_language())
 
+    # Caching run() in general seems too risky, but some callers know more
+    # about the programs they intend to run.
```

Comment:
Are there specific reasons you think it might be risky? To be honest it feels like we could just always cache it.

Reply:
User-supplied code can potentially do a lot of crazy things. Caching whether it successfully compiled/linked is relatively simple, caching whether it returns the same output when executed assumes that the code is intended to operate that way.

That was my reading, at least.

Reply:
Thinking more about this, there is an obvious case we don't want to cache: when the code is a File object. In that case we don't know if that file changed so we cannot use the cache. But... we cache cc.compiles() in that case it seems... that's probably a bug too then.

Reply:
Yea, the reason not to cache by default where along the lines Eli mentioned. The code could check that the stars and moons align in some magic way, for all we know. There's no way to inject potentially necessary dependencies to avoid problems. So it seems safest to just not cache by default.

And I think you have a point WRT the File issue. Perhaps we should just compute a hash of the file contents?

Reply:
I had been thinking to use mtime instead of filename. That's also what ninja would do if it was a build rule instead of a configure rule. :)

Reply:
I guess the way to fix this would be to fix the existing caching in a separate PR, and then use the infrastructure built for that here afterwards?

Reply:
I think they are totally unrelated, though? We could fix that separately in another PR and it won't hold up this one.

Reply:
> I think they are totally unrelated, though? We could fix that separately in another PR and it won't hold up this one.

I was thinking we'd fix this by introducing a function that helps build cache key fields for the File case. Which, as a bugfix, might need to be backpatched. So I thought it'd be best to introduce the helper for that first.

But:
> In fact, we always know that cached_run() is invoked with a str, not a FileOrString. :)

Right - seems we should just have cached_run() require a string? If at some later point files are required, it could be extended to support files at that point?

Reply:
Yes, that seems reasonable.

We cannot have the File part of FileOrString from meson internals, because we don't ship C sources inside the `mesonbuild` python module but always generate them from small inline code snippets. So it's only relevant for functions which take user input, and it's not obvious that there is any route whatsoever to making that cacheable here.

Cached compile/run operates on temporary files, we don't save every single .run() output on the off chance that after a reconfigure we might want to keep the old binary around so we can re-run it without recompiling it. And unless we commit to keeping all those built executables around to re-run them, there isn't really anything to cache.

If we *do* someday figure out a plan of action for that, I'm happy to update the function signature to accept File at that time. :)

Reply:
As I changed the signature to just accept code as a string, I think this is resolved now?

Reply:
Yes.